### PR TITLE
[BE-06] Add rating

### PR DESCRIPTION
**Pull request by pannoewo**
Add rate model
Create a contract api for rating recipes
Return rating in show recipes endpoint
AC:
User can vote on any recipe. (1-5)

### CODE REVIEW

File: app/controllers/api/v1/rates_controller.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+module Api
+  module V1
+    class RatesController < ApplicationController
+      before_action :set_recipe
+      
+    def create
```
Comment (DanielOldak) [inline app/controllers/api/v1/rates_controller.rb]: wrong indentation

File: app/controllers/api/v1/rates_controller.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        rating = Rate.new(ratings_params)
+        rating.recipe = @recipe
+ 
+        if rating.save
+          render json: rating
+        else
+          render json: { errors: rating.errors.full_messages }, status: :unprocessable_entity
+        end
+    end
```
Comment (DanielOldak) [inline app/controllers/api/v1/rates_controller.rb]: same as ```def```

File: app/controllers/api/v1/rates_controller.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+module Api
+  module V1
+    class RatesController < ApplicationController
+      before_action :set_recipe
+      
+    def create
+        rating = Rate.new(ratings_params)
+        rating.recipe = @recipe
```
Comment (DanielOldak) [inline app/controllers/api/v1/rates_controller.rb]:
```suggestion
        rate = @recipe.rate.new(rating_params)
```

File: spec/controllers/api/v1/rates_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        {
+          rate: { 
+           value: 4 
+          }, 
+          recipe_id: recipe.id
+        }
+        end
+      
+      it 'create a rating' do
+        expect { post_request}.to change(Rate, :count).by(1)
```
Comment (DanielOldak) [inline spec/controllers/api/v1/rates_controller_spec.rb]:
```suggestion
        expect { post_request }.to change(Rate, :count).by(1) 
```

File: spec/serializers/rate_serializer_spec.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+require 'rails_helper'
+
+describe RateSerializer do
+  subject {described_class.new(rate).to_h }
+  
+  let(:rate) { build :rate }
+  
+  it {is_expected.to include(value: rate.value)}
```
Comment (DanielOldak) [inline spec/serializers/rate_serializer_spec.rb]:
```suggestion
  it { is_expected.to include(value: rate.value) }
```

File: spec/serializers/rate_serializer_spec.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+require 'rails_helper'
+
+describe RateSerializer do
+  subject {described_class.new(rate).to_h }
```
Comment (DanielOldak) [inline spec/serializers/rate_serializer_spec.rb]:
```suggestion
  subject { described_class.new(rate).to_h }
```

File: app/controllers/api/v1/rates_controller.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    end
+
+      private
+
+      def ratings_params
+        params.require(:rate).permit(:value)
+      end
+
+      def set_recipe
+        @recipe = Recipe.find(params[:recipe_id])
```
Comment (DanielOldak) [inline app/controllers/api/v1/rates_controller.rb:23]: When somebody will pass recipe_id that doesn't exist ```.find``` will throw an exception. Please read about it and try to handle it 🙂

File: spec/controllers/api/v1/rates_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        {
+          rate: {
+            value: 4
+          },
+          recipe_id: recipe.id
+        }
+      end
+
+      it 'create a rating' do
+        expect { post_request }.to change(Rate, :count).by(1)
```
Comment (DanielOldak) [inline spec/controllers/api/v1/rates_controller_spec.rb:19]: 👌

File: spec/models/rate_spec.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
+require 'rails_helper'
+
+RSpec.describe Rate, type: :model do
+  describe 'associations' do
+    it { is_expected.to belong_to(:recipe) }
```
Comment (DanielOldak) [inline spec/models/rate_spec.rb:5]:
```suggestion
    it { is_expected.to belong_to(:recipe).dependent(:destroy) }
```

File: spec/serializers/recipe_show_serializer_spec.rb [hunk 10]
```diff
@@ -55,6 +55,19 @@
     end
   end
 
+
+  describe 'rates' do
```
Comment (grabomarcin) [inline spec/serializers/recipe_show_serializer_spec.rb]:
```suggestion
  describe '#avg_rate' do
```
Please add also spec for `#rate_count`.

File: spec/serializers/recipe_show_serializer_spec.rb [hunk 11]
```diff
@@ -68,6 +68,21 @@
...
+      it { is_expected.to include(rate_count: 2) }
+    end
+
+    context 'without rates' do
+      it { is_expected.to include(rate_count: 0) }
+    end
+ end
+ 
+ 
+
```
Comment (grabomarcin) [inline spec/serializers/recipe_show_serializer_spec.rb]:
```suggestion
```

**Review by grabomarcin — APPROVED** [review]

**Pull request merged**